remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RemotingSerializable.java

```diff
@@ -17,11 +17,17 @@
 package org.apache.rocketmq.remoting.protocol;
 
 import com.alibaba.fastjson.JSON;
+import com.alibaba.fastjson.serializer.SerializerFeature;
+
 import java.nio.charset.Charset;
 
 public abstract class RemotingSerializable {
     private final static Charset CHARSET_UTF8 = Charset.forName("UTF-8");
 
+    static {
```

Comment:
fastjson默认生成的key会产生不规范情况，如果只是重载方法，那样子就需要修改很多可能不规范的地方的调用，然后，改动量很大！现在只需要加这一个序列化属性，让rocketmq的所有remote调用结果全部是合法的json@我认为这是一个错误，而不算是一个优化了！

Reply:
这个我已经测试过了，这个remote模块会被admin和client都包含，所以这样子也能做到服务端和客户端的序列化属性的一致性，要不然可能会产生客户端和服务端的序列化属性不一致的情况！

Reply:
总体来说，我认为这是一个bug，而不是一个优化点，因为这个json数据，go无法解析，在线测试json格式也是错误的，然后jackson也无法解析，只能是fastjson自己产生，以及解析！这是不是就有点违背json原本跨语言的原则了！

Reply:
I agree it can be safely regarded as BUG for not generating standard-conforming **JSON**. But it would be necessary not to modify global behavior without notice. Consider this scenario, an application developer uses RocketMQ client and FastJSON at the same time. If our client quietly changes the default behavior, we may break his/her application or mess his data up totally in case he/she depends on the default behavior.

Further, IMO, RocketMQ-scope changes can be achieved through modifying RemotingSerializable.java#decode/encode only.

Reply:
Before modifying the behavior, we have to:
1, Make sure code can correctly deserialize JSON strings previously serialized without Compatible Feature configured;
2, Make sure clients without feature configured may compatibly parse standard JSON;
3, Verify JSON generated with feature configured can serialize and deserialize complex objects, say a map with non-string key.

Once these three compatible concerns are resolved, I am glad to +1 for this pull request.

Reply:
ok，我感觉这个最大的隐患可能是broker和namesrv是否需要同时设置SerializerFeature.WriteNonStringKeyAsString的问题，我在低版本的fastjson1.2.58好像，是必须同时设置的！如果最新版本的fastjson也需要同时设置，那就真得两边同时修改了！

Reply:
I suggest we adopt a test-driven development strategy for this issue.  Create 3 or more unit tests, which help us guarantee changes are always compatible with the community and discuss potential solutions where issues show up.

Reply:
测试了下fastjson的最新版本(1.2.76)是可以适配两种json格式的处理的，但是考虑到客户端的fastjson的版本是否也能够兼容和适配，存在风险性。
由于存在服务模块，namesrv，broker三者之间通过rpc的方式进行交流，所以，为了兼容，我想了如下方案来处理：
```
org.apache.rocketmq.remoting.protocol.RemotingCommand 这个类所有的请求命令类，在这个类里面有个属性：version，这个属性记录了当前请求的RocketMQ的版本号，那么就可以根据这个版本号来判断，toJSON()调用的时候是否需要设置：
SerializerFeature.WriteNonStringKeyAsString 属性，具体的伪代码如下：

if(version>= 设置了WriteNonStringKeyAsString版本号){
    return JSON.toJSON(obj,SerializerFeature.WriteNonStringKeyAsString);
}else{
    return JSON.toJSON(obj);
}
```
目前我只能想到这个兼容方案了，因为，可能存在客户端，namesrv和broker以及不同的broker版本号都可能不一致的情况，所以，你之前提到的三个问题，如果是通过我这种一棍子打死的方式来处理，确实很粗暴。那么正好就利用这个version属性来做这个版本的分割线判断！